In [1]:
import torch
a = torch.tensor([[1, 2, 3], [4, 5, 6]], dtype=torch.int32)
print(a)
print("\nShape (Kích thước):", a.shape)
print("Number of dimensions (Số chiều):", a.ndim)
print("Data type (Kiểu dữ liệu):", a.dtype)

tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)

Shape (Kích thước): torch.Size([2, 3])
Number of dimensions (Số chiều): 2
Data type (Kiểu dữ liệu): torch.int32


### Bài 1: Tạo Tensor cơ bản
* **Kích thước (Shape):** `torch.Size([2, 3])` (gồm 2 hàng và 3 cột).
* **Số chiều (ndim):** `2` (đây là ma trận 2 chiều).
* **Kiểu dữ liệu (dtype):** `torch.int32` (kiểu số nguyên 32-bit).

In [2]:
import torch
a = torch.rand(3, 4)
b = torch.ones(2, 3)
c = torch.zeros(2, 3)
d = torch.eye(4)
print(a)
print(b)
print(c)
print(d)
e = torch.full((5, 5), 10)
print("\nTensor e (5x5 toàn giá trị 10):\n", e)

tensor([[0.3910, 0.1238, 0.9371, 0.6665],
        [0.5814, 0.5553, 0.4465, 0.9359],
        [0.5479, 0.5155, 0.6107, 0.1012]])
tensor([[1., 1., 1.],
        [1., 1., 1.]])
tensor([[0., 0., 0.],
        [0., 0., 0.]])
tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])

Tensor e (5x5 toàn giá trị 10):
 tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


### Nhận xét & Giải thích - Bài 2
* **Nhận xét sự khác nhau giữa `torch.rand()` và `torch.randn()`:**
  * `torch.rand()`: Sinh các số ngẫu nhiên theo **phân bố đều (Uniform Distribution)** trong khoảng $[0, 1)$.
  * `torch.randn()`: Sinh các số ngẫu nhiên theo **phân bố chuẩn (Standard Normal / Gaussian Distribution)** với Trung bình (Mean) = 0 và Độ lệch chuẩn (Std) = 1.

In [12]:
import torch

a = torch.zeros(2, 3, 4)

print("Tensor:")
print(a)
print("Shape:", a.shape)
print("Size:", a.size())
print("Number of dimensions:", a.ndim)
print("Data type:", a.dtype)

Tensor:
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
Shape: torch.Size([2, 3, 4])
Size: torch.Size([2, 3, 4])
Number of dimensions: 3
Data type: torch.float32


### Nhận xét & Giải thích - Bài 3
* **Giải thích ý nghĩa `torch.Size([2, 3, 4])`:**
  * Thể hiện đây là một Tensor 3 chiều (3D Tensor).
  * Chiều thứ nhất (index 0): Có kích thước 2 (chứa 2 ma trận 2D).
  * Chiều thứ hai (index 1): Có kích thước 3 (mỗi ma trận gồm 3 hàng).
  * Chiều thứ ba (index 2): Có kích thước 4 (mỗi hàng gồm 4 cột).
  * Tổng số phần tử trong Tensor là $2 \times 3 \times 4 = 24$ phần tử.

In [3]:
import torch

a = torch.randn(3, 4, 5)
print("Original tensor shape:", a.shape)
print("\nTensor a[1] shape:", a[1].shape)
print("Tensor a[1:, 2:4] shape:", a[1:, 2:4].shape)
two_last_elements = a[:, :, -2:]
print("\nHai phần tử cuối cùng trên chiều thứ ba có shape:", two_last_elements.shape)

Original tensor shape: torch.Size([3, 4, 5])

Tensor a[1] shape: torch.Size([4, 5])
Tensor a[1:, 2:4] shape: torch.Size([2, 2, 5])

Hai phần tử cuối cùng trên chiều thứ ba có shape: torch.Size([3, 4, 2])


### Nhận xét & Giải thích - Bài 4
* **Shape của `a[1]`:** `torch.Size([4, 5])` (do đã chọn chỉ số 1 ở chiều thứ 0 nên chiều 0 bị tiêu biến).
* **Shape của `a[1:, 2:4]`:** `torch.Size([2, 2, 5])`
  * Chiều 0: Lấy từ index 1 đến hết $\rightarrow$ 2 phần tử (index 1 và 2).
  * Chiều 1: Lấy từ index 2 đến trước 4 $\rightarrow$ 2 phần tử (index 2 và 3).
  * Chiều 2: Lấy toàn bộ $\rightarrow$ 5 phần tử.

In [ ]:
import torch

a = torch.randn(3, 4)
print("Original:")
print(a)

print("\nFlatten:")
print(a.ravel())

print("\nReshape to 3 x 2 x 2:")
print(a.reshape(3, 2, 2))

a_2x6 = a.reshape(2, 6)
print("\nReshape to 2 x 6:\n", a_2x6)

Original:
tensor([[-0.3586,  0.3817, -0.0383,  0.3297],
        [-1.6639, -0.3317,  0.3858,  0.4139],
        [-1.6691,  0.8219,  1.9973, -1.1845]])

Flatten:
tensor([-0.3586,  0.3817, -0.0383,  0.3297, -1.6639, -0.3317,  0.3858,  0.4139,
        -1.6691,  0.8219,  1.9973, -1.1845])

Reshape to 3 x 2 x 2:
tensor([[[-0.3586,  0.3817],
         [-0.0383,  0.3297]],

        [[-1.6639, -0.3317],
         [ 0.3858,  0.4139]],

        [[-1.6691,  0.8219],
         [ 1.9973, -1.1845]]])

Reshape thành 2 x 6:
 tensor([[-0.3586,  0.3817, -0.0383,  0.3297, -1.6639, -0.3317],
        [ 0.3858,  0.4139, -1.6691,  0.8219,  1.9973, -1.1845]])


### Nhận xét & Giải thích - Bài 5
* **Kiểm tra số phần tử trước và sau khi reshape:** Số phần tử luôn **không đổi** và bằng 12 phần tử ($3 \times 4 = 12$ và $3 \times 2 \times 2 = 12$). Nguyên tắc khi reshape là tổng số phần tử ban đầu phải đúng bằng tích kích thước các chiều mới.

In [18]:
import torch

a = torch.randn(2, 3)
b = torch.randn(2, 3)
print("a:") 
print(a) 
print("b:") 
print(b)

a:
tensor([[-0.8615, -2.1852,  0.4868],
        [-0.6900,  0.9494,  0.6704]])
b:
tensor([[-1.1083,  1.5148,  0.1775],
        [-0.2853, -1.4812, -1.0101]])


### Nhận xét & Giải thích - Bài 6
* **Phân biệt `a * b` với `a @ b.T`:**
  * `a * b`: Là phép nhân **Element-wise** (nhân từng phần tử tương ứng). Cả 2 tensor phải cùng shape $(2, 3)$, kết quả ra shape $(2, 3)$.
  * `a @ b.T`: Là phép **nhân ma trận đại số tuyến tính**. Ma trận $a$ kích thước $(2, 3)$ nhân với ma trận chuyển vị $b^T$ kích thước $(3, 2)$.
* **Shape của `a @ b.T`:** `torch.Size([2, 2])`.

In [20]:
import torch

a = torch.randn(3, 4)
print("Tensor a:\n", a)

print("\n--- Khi dim=0 (Tính theo cột) ---")
print("Mean:", torch.mean(a, dim=0))
print("Std:", torch.std(a, dim=0))

print("\n--- Khi dim=1 (Tính theo hàng) ---")
print("Mean:", torch.mean(a, dim=1))
print("Std:", torch.std(a, dim=1))
print("Cumulative sum:\n", torch.cumsum(a, dim=1))

Tensor a:
 tensor([[ 1.2707,  0.4265, -1.4125, -1.2308],
        [-1.4117, -0.2603, -1.2731,  0.0120],
        [ 0.4301,  0.2522, -0.4715, -1.2214]])

--- Khi dim=0 (Tính theo cột) ---
Mean: tensor([ 0.0964,  0.1395, -1.0524, -0.8134])
Std: tensor([1.3720, 0.3570, 0.5078, 0.7149])

--- Khi dim=1 (Tính theo hàng) ---
Mean: tensor([-0.2365, -0.7333, -0.2527])
Std: tensor([1.3016, 0.7144, 0.7544])
Cumulative sum:
 tensor([[ 1.2707,  1.6972,  0.2848, -0.9460],
        [-1.4117, -1.6719, -2.9451, -2.9330],
        [ 0.4301,  0.6823,  0.2108, -1.0106]])


### Nhận xét & Giải thích - Bài 7
* **Nhận xét sự khác nhau giữa `dim=0` và `dim=1`:**
  * `dim=0`: Thực hiện tính toán **theo từng cột** (tính dọc theo các hàng). Trả về tensor có 4 phần tử (tương ứng 4 cột).
  * `dim=1`: Thực hiện tính toán **theo từng hàng** (tính ngang qua các cột). Trả về tensor có 3 phần tử (tương ứng 3 hàng).

In [21]:
import torch

x1 = torch.tensor(3.6, requires_grad=True)
y1 = x1 * x1
y1.backward()
print("Hàm y = x^2:")
print("x.grad (PyTorch):", x1.grad)

x2 = torch.tensor(3.6, requires_grad=True)
y2 = 3 * x2**2 + 2 * x2 + 1
y2.backward()
print("\nHàm y = 3*x^2 + 2*x + 1:")
print("x.grad (PyTorch):", x2.grad)

Hàm y = x^2:
x.grad (PyTorch): tensor(7.2000)

Hàm y = 3*x^2 + 2*x + 1:
x.grad (PyTorch): tensor(23.6000)


### Nhận xét & Giải thích - Bài 8
* **Tính thủ công đạo hàm $y = x^2$ tại $x = 3.6$:**
  * Đạo hàm lý thuyết: $y' = \frac{dy}{dx} = 2x$.
  * Tại $x = 3.6 \Rightarrow y' = 2 \times 3.6 = 7.2$.
  * **So sánh:** Kết quả tính thủ công ($7.2$) trùng khớp hoàn toàn với `x.grad` từ PyTorch ($7.2000$).
* **Kiểm tra hàm mới $y = 3x^2 + 2x + 1$ tại $x = 3.6$:**
  * Đạo hàm lý thuyết: $y' = 6x + 2$.
  * Tại $x = 3.6 \Rightarrow y' = 6(3.6) + 2 = 21.6 + 2 = 23.6$. Kết quả PyTorch trả về đúng $23.6$.

In [22]:
import numpy as np
import torch

polynomial = np.poly1d([1, 2, 3])
N = 20
X = np.random.randn(N, 1)
Y = polynomial(X)

XX = np.hstack((X * X, X, np.ones_like(X)))
x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)
w = torch.randn(3, 1, requires_grad=True)

optimizer = torch.optim.NAdam([w], lr=0.01)

for epoch in range(1000):
    y_pred = x @ w
    mse = torch.mean(torch.square(y_pred - y))
    optimizer.zero_grad()
    mse.backward()
    optimizer.step()

print("Learned coefficients (Hệ số học được):\n", w.detach().numpy().flatten())
print("Target coefficients (Hệ số mục tiêu): [1. 2. 3.]")

Learned coefficients (Hệ số học được):
 [1.3045478 2.218541  2.6098561]
Target coefficients (Hệ số mục tiêu): [1. 2. 3.]


### Nhận xét & Giải thích - Bài 9
* **So sánh 3 hệ số học được:** Trọng số `w` ban đầu khởi tạo ngẫu nhiên, sau 1000 epoch huấn luyện đã hội tụ về rất sát với hệ số mục tiêu thực tế `[1, 2, 3]`.
* **Giải thích vai trò của các hàm:**
  * `optimizer.zero_grad()`: Xóa sạch các giá trị gradient đã lưu ở bước trước. Nếu không reset, gradient sẽ bị cộng dồn khiến quá trình cập nhật bị sai lệch.
  * `mse.backward()`: Thực hiện lan truyền ngược (backpropagation) để tự động tính gradient của hàm tổn thất (MSE) theo các trọng số `w` và lưu vào `w.grad`.
  * `optimizer.step()`: Cập nhật trọng số `w` dựa trên giá trị gradient vừa tính được theo thuật toán tối ưu NAdam.

In [23]:
import random
import torch

EPOCHS = 2000
A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)

optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)

for _ in range(EPOCHS):
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2
    sqerr = y1*y1 + y2*y2 + y3*y3 + y4*y4
    
    optimizer.zero_grad()
    sqerr.backward()
    optimizer.step()

print(f"A = {A.item():.4f}, B = {B.item():.4f}, C = {C.item():.4f}, D = {D.item():.4f}")
print("Kiểm tra A + B (Mục tiêu 9):", (A + B).item())
print("Kiểm tra C - D (Mục tiêu 1):", (C - D).item())
print("Kiểm tra A + C (Mục tiêu 8):", (A + C).item())
print("Kiểm tra B - D (Mục tiêu 2):", (B - D).item())

A = 4.6329, B = 4.3670, C = 3.3670, D = 2.3670
Kiểm tra A + B (Mục tiêu 9): 8.9999418258667
Kiểm tra C - D (Mục tiêu 1): 1.0000219345092773
Kiểm tra A + C (Mục tiêu 8): 7.9999799728393555
Kiểm tra B - D (Mục tiêu 2): 1.9999840259552002


### Nhận xét & Giải thích - Bài 10
* **Kiểm tra nghiệm:** Thay bộ nghiệm $A, B, C, D$ tìm được vào 4 phương trình ban đầu cho kết quả sai số xấp xỉ 0.
* **Nhận xét vì sao bài toán có nhiều nghiệm:**
  * Ta có 4 phương trình: (1) $A+B=9 \Rightarrow B=9-A$, (2) $C-D=1 \Rightarrow D=C-1$, (3) $A+C=8 \Rightarrow C=8-A$, (4) $B-D=2$.
  * Thế các ẩn theo $A$ vào (4): $(9-A) - (8-A-1) = 2 \Leftrightarrow 2 = 2$ (luôn đúng với mọi $A$).
  * Phương trình (4) phụ thuộc tuyến tính vào 3 phương trình còn lại, nên hệ chỉ có 3 phương trình độc lập cho 4 ẩn $\Rightarrow$ Tồn tại **vô số nghiệm**. Mô hình sẽ hội tụ về nghiệm nào tùy thuộc vào giá trị khởi tạo ban đầu.

In [24]:
import torch
import numpy as np

print("=== Câu 1: Reshape Tensor X (24 phần tử) ===")
X = torch.arange(24)
print("Reshape 4x6:\n", X.reshape(4, 6))
print("Reshape 2x12:\n", X.reshape(2, 12))
print("Reshape 2x3x4:\n", X.reshape(2, 3, 4))

print("\n=== Câu 2: Các phép toán trên A, B (3x4) ===")
A = torch.randn(3, 4)
B = torch.randn(3, 4)
print("A + B:\n", A + B)
print("A * B:\n", A * B)
print("A / B:\n", A / B)
print("Trung bình A theo cột:", torch.mean(A, dim=0))
print("Độ lệch chuẩn A theo cột:", torch.std(A, dim=0))

print("\n=== Câu 3: Autograd cho y = 2x^2 + 5x + 3 tại x = 2 ===")
x = torch.tensor(2.0, requires_grad=True)
y = 2 * x**2 + 5 * x + 3
y.backward()
print("dy/dx tại x=2 (PyTorch Autograd):", x.grad.item())
print("dy/dx tính thủ công (4*2 + 5):", 4 * 2 + 5)

print("\n=== Câu 4: Tìm hệ số cho y = 2x^2 - 3x + 5 ===")
poly = np.poly1d([2, -3, 5])
N = 30
X_np = np.random.randn(N, 1)
Y_np = poly(X_np)

XX = np.hstack((X_np * X_np, X_np, np.ones_like(X_np)))
x_ten = torch.tensor(XX, dtype=torch.float32)
y_ten = torch.tensor(Y_np, dtype=torch.float32)
w = torch.randn(3, 1, requires_grad=True)

opt = torch.optim.NAdam([w], lr=0.01)
for epoch in range(1500):
    pred = x_ten @ w
    loss = torch.mean(torch.square(pred - y_ten))
    opt.zero_grad()
    loss.backward()
    opt.step()

print("Hệ số học được [a, b, c]:", w.detach().numpy().flatten())
print("Hệ số thực tế: [ 2. -3.  5.]")

=== Câu 1: Reshape Tensor X (24 phần tử) ===
Reshape 4x6:
 tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]])
Reshape 2x12:
 tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]])
Reshape 2x3x4:
 tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])

=== Câu 2: Các phép toán trên A, B (3x4) ===
A + B:
 tensor([[-1.4728,  0.3236,  2.1695,  0.9298],
        [-1.1862, -0.1122, -2.1222, -2.2584],
        [-2.5946,  0.1388, -0.3423, -2.7211]])
A * B:
 tensor([[-0.7708,  0.0261,  1.0397,  0.1552],
        [-0.0761, -0.0079,  0.4318,  1.1859],
        [ 1.3209, -0.0213,  0.0249,  1.7862]])
A / B:
 tensor([[ -0.2176,   1.1239,   0.4911,   3.2643],
        [-20.4396,  -0.3029,   0.1203,   1.7193],
        [  0.3663,  -0.3989,   0.4394,   0

### Nhận xét & Giải thích - Bài 11
1. **Reshape:** Chuyển đổi thành công Tensor 24 phần tử sang các kích thước $4 \times 6$, $2 \times 12$ và $2 \times 3 \times 4$.
2. **Phép toán ma trận:** Thực hiện đúng các phép tính phần tử (cộng, nhân, chia) và tính giá trị trung bình, độ lệch chuẩn theo từng cột (`dim=0`).
3. **Kiểm tra Autograd:** Hàm $y = 2x^2 + 5x + 3 \Rightarrow y' = 4x + 5$. Tại $x = 2$, $y' = 13$. Autograd trả về kết quả 13 trùng khớp với tính thủ công.
4. **Gradient Descent đa thức mới:** Hệ số học được hội tụ về cực kỳ sát với các hệ số thực tế $[2, -3, 5]$.